File: FrequentAlgorithm.java
import java.util.HashMap;
import java.util.Map;

public class FrequentAlgorithm {

    static final int k = 4;
    static final int monitorCount = k - 1;

    public static void main(String[] args) {
        String[] sequence = {"L", "K", "M", "P", "L", "K", "K", "M", "P", "K", "K", "P", "M", "L", "P"};
        Map<String, Integer> monitors = new HashMap<>();

        for (String item : sequence) {
            if (monitors.containsKey(item)) {
                monitors.put(item, monitors.get(item) + 1);
            } else if (monitors.size() < monitorCount) {
                monitors.put(item, 1);
            } else {
                decrementMonitors(monitors);
            }
            printMonitors(monitors);
        }
    }

    private static void decrementMonitors(Map<String, Integer> monitors) {
        monitors.entrySet().removeIf(entry -> entry.getValue() == 1);
        monitors.replaceAll((key, value) -> value - 1);
    }

    private static void printMonitors(Map<String, Integer> monitors) {
        System.out.println("Monitors: " + monitors);
    }
}

File: FrequentAlgorithm.py
def decrement_monitors(monitors):
    keys_to_remove = []
    for key in monitors:
        monitors[key] -= 1
        if monitors[key] == 0:
            keys_to_remove.append(key)
    
    for key in keys_to_remove:
        del monitors[key]

def frequent_algorithm(sequence, k):
    monitor_count = k - 1
    monitors = {}

    for item in sequence:
        if item in monitors:
            monitors[item] += 1
        elif len(monitors) < monitor_count:
            monitors[item] = 1
        else:
            decrement_monitors(monitors)
        print_monitors(monitors)

def print_monitors(monitors):
    print("Monitors:", monitors)

sequence = ['L', 'K', 'M', 'P', 'L', 'K', 'K', 'M', 'P', 'K', 'K', 'P', 'M', 'L', 'P']
k = 4
frequent_algorithm(sequence, k)
